pyogrio/geopandas.py

```diff
@@ -313,7 +313,8 @@ def write_dataframe(
 
         # If there is data, infer layer geometry type + promote_to_multi
         if not df.empty:
-            has_z = geometry.has_z.any()
+            # Empty geometries sometimes report as Z incorrectly, so ignore them
+            has_z = geometry[~geometry.is_empty].has_z.any()
```

Comment:
Since the error from GDAL for FlatGeobuf on mixed dimensions is not very easy to interpret, I wonder if we can't catch it sooner - at least for the GeoDataFrame case?

Perhaps something like:

```Python
ix = geometry[~geometry.is_empty].has_z
has_z = ix.any()

if driver == "FlatGeobuf" and has_z and not ix.all():
    raise DataSourceError("FlatGeobuf does not support mixed 2D and 3D coordinates")
```

This may also be a place to generalize via a `DRIVERS_NO_MIXED_DIMENSIONS` similar to the way `DRIVERS_NO_MIXED_SINGLE_MULTI` is used.

Reply:
I added a cleaner error... but fgb seems to be a can of worms: a new problem arose: #226